keps/sig-node/24-apparmor/README.md

```diff
@@ -0,0 +1,593 @@
+
+The promotion of AppArmor to GA would deprecate the beta annotations as described in the
+[Version Skew Strategy](#version-skew-strategy).
+
+### Monitoring requirements
+
+###### How can an operator determine if the feature is in use by workloads?
+
+The feature is built into the kubelet and api server components. No metric is
+planned at this moment. The way to determine usage is by checking whether the
+pods/containers have a AppArmorProfile set.
+
+###### How can someone using this feature know that it is working for their instance?
+
+The AppArmor enforcement status is not directly surfaced by Kubernetes, but is visible through the
+linux proc API:
+
+```sh
+$ cat /proc/1/attr/current
+k8s-apparmor-example-deny-write (enforce)
+```
+
+###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?
+
+N/A
+
+###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?
+
+N/A
+
+###### Are there any missing metrics that would be useful to have to improve observability of this feature?
+
+N/A
+
+### Dependencies
+
+###### Does this feature depend on any specific services running in the cluster?
+
+No
+
+### Scalability
+
+###### Will enabling / using this feature result in any new API calls?
+
+No
+
+###### Will enabling / using this feature result in introducing new API types?
+
+No
+
+###### Will enabling / using this feature result in any new calls to the cloud provider?
+
+No
+
+###### Will enabling / using this feature result in increasing size or count of the existing API objects?
+
+Negligible increase in Pod object size, and any objects embedding a PodSpec.
+
+###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?
+
+No
+
+###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?
+
+No
+
+### Troubleshooting
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+No impact to running workloads.
+
+###### What are other known failure modes?
+
+No impact is being foreseen to running workloads based on the nature of
+changes brought by this KEP.
+
+Although some general errors and failures can be seen on [Failure and Fallback
+Strategy](#failure-and-fallback-strategy).
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+N/A
+
+## Implementation History
+
+- 2016-07-25: AppArmor design proposal
+- 2016-09-26: AppArmor beta release with v1.4
+- 2020-01-10: Initial KEP
+
+## Drawbacks
+
+Promoting AppArmor as-is to GA may be seen as "blessing" the current
+functionality, and make it harder to make some of the enhancements listed under
+[Non-Goals](#non-goals). Since the current behavior is unguarded, I think we
+already need to treat the behavior as GA.
+
+## Alternatives
+
+N/A
```

Comment:
My opinion: we could have a beta release that switches AppArmor to use API fields for Pod, and do the graduation to GA in the minor release after that.

Jumping in one go gives people no option to opt out of a field that wasn't there at all in the previous release.


I suggest we document this more careful approach so that it's clear what we ruled out and why.

Reply:
Is the only difference with your suggestion that we'd make all the changes proposed here, but the feature gate would still be labeled as Beta? Then next release the only change is to update the feature gate status? Why does that separation matter?

Reply:
You can disable beta feature gates, and you can't disable GA feature gates. It'd give people a chance to discover whether the new field is problematic and opt out, without forcing them to downgrade if it is.